Ci: multi branch build

a-gave opened the pull request:
These changes are made:
- adapt packages versioning scheme to apk.
- split the main ci build to use the new package manager apk, while keeping the default feed for opkg
- add multi branch build for shared-state-async
- update README

Details:
chore: add include libremesh.mk in more packages
  modify various packages to use libremesh.mk
  drop the package lua-argparse since it is available in openwrt feeds.

  modify PKG_VERSION to follow apk semantics:
    use by default $(GIT_COMMIT_DATE)~$(GIT_COMMIT_TSTAMP)
    like 2024.11.16~1731788517
    if a specific PKG_VERSION is not defined  
  From openwrt/packages:
  Preparations to switch to APK package manager https://github.com/openwrt/packages/issues/23706

ci: update feeds scheme to build C++ packages for 3 latest openwrt branches
  the new scheme make these packages available at
  https://github.com/libremesh/lime-feed/
  referenced by apk in /etc/apk/repositories as
  @libremesh_arch_packages https://feed.libremesh.org/arch_packages/master/openwrt_main/x86_64/packages.adb
  
ci: disable sign packages in main openwrt branch
  not yet supported, there is an active pull 
  Support sign apk feed https://github.com/openwrt/gh-action-sdk/pull/46

  to install packages use "apk add --allow-untrusted lime-docs"

lime-system: add apk option in 92_add-lime-repos
chore: update README.md with instruction for apk

a-gave commented:
push un-pushed fixes: external_repositories was set to fork.

update apk repositories file to reflect latest changes in openwrt,
creating a /etc/apk/repositories.d/limefeeds.list for lime-packages:
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=e031dab93ec3a066880ff25e0f42d6bef1e3001a
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=a6e98a810d4d831caf9f18536701c433b9e57fc8

ilario commented:
Wow this is a huge amount of work!
Do you think we should include this in the upcoming 2024.1 release or is it ok to leave it for a next release?
Is this also related to https://github.com/libremesh/lime-packages/issues/1135 ?

a-gave commented:
I'm doing some fixes

the new workflow build.yml should be added soon because next push will use openwrt-main with apk, thus not building almost every lime-packages because of incorrect PKG_VERSION

Yes it is related to because to use shared-state-async on new builds based on 24.10.x or 23.05.x you should use the feed libremesh_arch_packages if you look for a precompiled one

a-gave commented:
it should be now ready 
the workflow multi-arch-build was failing on publishing packages but it should be ok now, i'm testing with all archs

a-gave commented:
I confirm that the fail on publishing is fixed since packages are now produced in this fork
https://github.com/a-gave/lime-feed/tree/gh-pages/arch_packages/master/

I updated the description adding more notes on this.

a-gave commented:
Just noticed the fail of job build packages with Merge pull request #1133 from henmohr/issue/1077 
https://github.com/libremesh/lime-packages/actions/runs/11881297358
and add another note on the description mentioning the update of openwrt-sdk-action

javierbrk commented:
Amazing job !! 
I did a similar task a couple of weeks ago, while building images for librerouter sadly I run out of space after switching to the new openWrt version. The good news is that the kernel can be reduced in size and i managed to get some usable images. 
 
There are some options that may be useful to make the kernel lose some weight. All of them related to the SMALL_FLASH flag. 
Here is a list of relevant flags to take into account. I 'll mention them just in case you need them.

config KERNEL_MPTCP
 bool "Multi-Path TCP support"
 default y if !SMALL_FLASH
 help
    Select this option to enable support for Multi-Path TCP.
    **Increases the compressed kernel size by ~214kB (as of Linux 6.6).**
    

config KERNEL_KALLSYMS
 bool "Compile the kernel with symbol table information"
 default y if !SMALL_FLASH
 help
   This will give you more information in stack traces from kernel oopses.
   
config KERNEL_DEBUG_INFO
 bool "Compile the kernel with debug information"
 default y if !SMALL_FLASH
 select KERNEL_DEBUG_KERNEL
 help
   This will compile your kernel and modules with debug information.

config KERNEL_ELF_CORE
 bool "Enable process core dump support"
 select KERNEL_COREDUMP
 default y if !SMALL_FLASH
 
config KERNEL_KEYS
 bool "Enable kernel access key retention support"
 default !SMALL_FLASH
 
config KERNEL_CGROUPS
 bool "Enable kernel cgroups"
 default y if !SMALL_FLASH 
 
config KERNEL_NAMESPACES
 bool "Enable kernel namespaces"
 default y if !SMALL_FLASH
 
config KERNEL_DEVPTS_MULTIPLE_INSTANCES
 bool "Support multiple instances of devpts"
 default y if !SMALL_FLASH
 help
   Enable support for multiple instances of devpts filesystem.
   If you want to have isolated PTY namespaces (eg: in containers),
   say Y here. Otherwise, say N. If enabled, each mount of devpts
   filesystem with the '-o newinstance' option will create an
   independent PTY namespace.

config KERNEL_SECCOMP_FILTER
 bool
 default y if !SMALL_FLASH  

config KERNEL_SECCOMP
 bool "Enable seccomp support"
  depends on !(TARGET_uml)
  select KERNEL_SECCOMP_FILTER
  default y if !SMALL_FLASH
  help
    Build kernel with support for seccomp.


config KERNEL_NET_L3_MASTER_DEV
 bool "L3 Master device support"
 default y if !SMALL_FLASH
 help
   This module provides glue between core networking code and device
   drivers to support L3 master devices like VRF.
     **Increases the compressed kernel size by ~4kB (as of Linux 6.6).**



menu "Filesystem ACL and attr support options"
 config USE_FS_ACL_ATTR
  bool "Use filesystem ACL and attr support by default"
  default y if !SMALL_FLASH
  help
    Make using ACLs (e.g. POSIX ACL, NFSv4 ACL) the default
    for kernel and packages, except old NFS.
    Also enable userspace extended attribute support
    by default.  (OpenWrt already has an expection it will be
    present in the kernel).
    
config KERNEL_EXT4_FS_SECURITY
 bool "Ext4 Security Labels"
 default y if !SMALL_FLASH

config KERNEL_F2FS_FS_SECURITY
 bool "F2FS Security Labels"
 default y if !SMALL_FLASH

config KERNEL_UBIFS_FS_SECURITY
 bool "UBIFS Security Labels"
 default y if !SMALL_FLASH

config KERNEL_JFFS2_FS_SECURITY
 bool "JFFS2 Security Labels"
 default y if !SMALL_FLASH
 
config KERNEL_FHANDLE
 bool "Compile the kernel with support for fhandle syscalls"
 default y if !SMALL_FLASH

config KERNEL_FANOTIFY
 bool "Compile the kernel with modern file notification support"
 default y if !SMALL_FLASH

config KERNEL_IPV6_SEG6_LWTUNNEL
  bool "Enable support for lightweight tunnels"
  default y if !SMALL_FLASH
  help
    Using lwtunnel (needed for IPv6 segment routing) requires ip-full package.

a-gave commented:
> There are some options that may be useful to make the kernel lose some weight. All of them related to the SMALL_FLASH flag. Here is a list of relevant flags to take into account. I 'll mention them just in case you need them.

Thank you a lot for this information, I will try it for sure
Apparently there is since 2016: https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=commit&s=SMALL_FLASH
Maybe the wiki lacks a page explicitly mentioning it (other than changelogs), I had never noticed it anyway: https://openwrt.org/toh/start?do=search&id=toh%3Astart&sf=1&q=%22small_flash%22

ilario commented:
Hi! Just realized that in [OpenWrt 24.10 APK is not yet supported](https://openwrt.org/releases/24.10/notes-24.10.0#general_changes) (but it is in the main branch).

So, if we merge this pull request, will the compilation for 23.05 and 24.10 break? (I don't think so but a confirmation would be good @a-gave)

a-gave commented:
This and https://github.com/libremesh/network-profiles/pull/96 can now be merged

a-gave commented:
Marking this again as draft as 
- it should split the main feed per openwrt-branch to prevent issue related to changes in package names #1190
- the action openwrt-sdk-action has been updated with support for signed packages with apk

a-gave commented:
this is now mergeable, (@ilario , @G10h4ck , @javierbrk)
these are the latest changes:

## lime-app and libremesh.mk
I reverted the changes to include `libremesh.mk` in the package lime-app.    
Since by now `libremesh.mk` doesn't handle correctly packages that download stuff:     
like tar.gz archives (shared-state-async), another git repo (lime-docs), etc.

## shared-state-async
removed the workaround used to compile `shared-state-async` (to compile a package that select it)
apparently [0] the Makefile should contain either 
- a PKG_HASH (with the sha256sum of a downloaded tar.gz of a project, like lime-app)
- or in this case a PKG_MIRROR_HASH with the sha256sum of the git archive

adding this PKG_MIRROR_HASH using the procedure described here [0] (running `download` and then `check` with `FIXUP=1`) was not enough.     
The PKG_VERSION should stay fixed and it was instead 
- changing always the date to the 'actual date of build', generating every day a different hash that differs from the PKG_MIRROR_HASH regitered the day before.
- changing in the usage of dashes or dots in the GIT_COMMIT_DATE between openwrt-23.05 and subsequent releases, that also was generating different hash between openwrt branchs
- changing in the usage of `tar.zst` and `tar.xz` between openwrt-23.05 and subsequent releases

So I converted the PKG_VERSION to be i.e 2024.05.13-db58e3d-r1.apk      
with `r1` taken from `PKG_RELEASE` that is used as reference to the package 'wrapper' inside the repo lime-packages.
Added for now a double PKG_MIRROR_HASH for compatibility with openwrt-23.05

## change to more meaningful paths for feeds
I restructured the paths of the feeds to make them more meaningful.    
And split also packages that are compiled only for x86_64 for the 3 latest branch of openwrt     
This is done to prevent errors with sporadic package name changes, so to close also #1190

The new feeds to had in a device or in an imagebuilder to download the precompiled packages are now:
```
cat /etc/apk/repositories.d/limefeeds.list
https://feed.libremesh.org/master/openwrt-main/x86_64/packages.adb
https://feed.libremesh.org/master/openwrt-main/mips_24kc/packages.adb
```
and for previous opkg based releases:
```
cat /etc/opkg/limefeeds.conf
src/gz libremesh https://feed.libremesh.org/master/openwrt-23.05/x86_64
src/gz libremesh_arch_packages https://feed.libremesh.org/master/openwrt-23.05/mips_24kc
```
To test this precompiled packages you can replace feed.libremesh.org with these urls taking them from my repo `lime-feed`:

to use openwrt SNAPSHOT      
https://a-gave.github.io/lime-feed/master/openwrt-main/x86_64/packages.adb     
https://a-gave.github.io/lime-feed/master/openwrt-main/mips_24kc/packages.adb

to use openwrt 24.10-SNAPSHOT or v24.10.1 v24.10.2      
https://a-gave.github.io/lime-feed/master/openwrt-24.10/x86_64     
https://a-gave.github.io/lime-feed/master/openwrt-24.10/mips_24kc

i.e. equivalent to use openwrt SNAPSHOT      
https://raw.githubusercontent.com/a-gave/lime-feed/refs/heads/gh-pages/master/openwrt-main/x86_64/packages.adb     
https://raw.githubusercontent.com/a-gave/lime-feed/refs/heads/gh-pages/master/openwrt-main/mips_24kc/packages.adb


## new apk key
Added a new apk private key in the repo secrets as PRIVATE_KEY

[0] https://openwrt.org/docs/guide-developer/packages?s[]=pkg&s[]=mirror&s[]=hash#use_source_repository

a-gave commented on packages/bmx7-mdns/Makefile:
```diff
@@ -52,26 +48,4 @@ define Package/$(PKG_NAME)/description
-endef
-
-$(eval $(call BuildPackage,bmx7-mdns))
+$(eval $(call BuildPackage,$(PKG_NAME)))
```
fixed

a-gave commented on packages/random-numgen/Makefile:
```diff
@@ -2,17 +2,15 @@
-PKG_NAME:=random-numgen
-PKG_VERSION:=0.1
+PKG_REALVERSION:=0.1
+PKG_VERSION:=0.0.1
```
removed the PKG_REALVERSION variable

a-gave commented on packages/random-numgen/Makefile:
```diff
@@ -26,15 +24,4 @@ define Package/random-numgen/description
-endef
-
-$(eval $(call BuildPackage,random-numgen))
+$(eval $(call BuildPackage,$(PKG_NAME)))
```
fixed

a-gave commented on packages/shared-state-async/Makefile:
```diff
@@ -17,22 +17,32 @@
 # along with this program. If not, see <https://www.gnu.org/licenses/>
 #
-# SPDX-License-Identifier: AGPL-3.0-only
+# SPDX-License-Identifier: AGPL-3.0-only 
```
fixed

a-gave commented:
> Thanks for the contribution, overall seems quite good, I don't have strong background on github actions so take my comments on that part with a pinch of salt. Please address the other minor suggestions I made inline. @aparcar do you think this is the right direction to move forward?

Thanks for the review, pushed the minor fixes!

a-gave commented:
Merging to close #1190 and #1135 and to allow further updates of the firmware-selector to do builds with the imagebuilder for SNAPSHOT, openwrt-24.10 and openwrt-23.05, useful to test libremesh between releases.
